BigQuery: Load and dump schema from/to a file

jeremywadsack opened the pull request:
Addresses #1919 

This pull request adds the ability to load a BigQuery table schema from a JSON file (or other JSON source). And to write the schema of an existing table schema to a file. The file format is the same as the [CLI JSON file](https://cloud.google.com/bigquery/docs/schemas#using_a_json_schema_file).

@quartzmo I need some help with the doctest: I can't figure out how to stub `File.write` with minitest mocks for the scope of the doctest.

@blowmage In your [gist](https://gist.github.com/blowmage/e86d4b8d5b07b0d496d1234064b49e7f) you suggested class methods `Schema::load` and `Schema::dump`. Those are easy enough to implement, but while trying to document them, I realized that I have no idea when you would actually use that. Can you provide a suggestion? (It seems to me that Schema is _always_ used within the context of a table in which case there's an instance of it already.)

quartzmo commented:
@jeremywadsack Great to see this PR! I'll look through it soon.

> I need some help with the doctest: I can't figure out how to stub File.write with minitest mocks for the scope of the doctest.

I think we just monkeypatch (rather than mock) for doctest. Take a look at [`File.read` in the Storage doctest helper](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-storage/support/doctest_helper.rb#L18-L28).

jeremywadsack commented:
@quartzmo Thanks, updated the doctest.

CI failures are in the DataTransfer client. I think they are unrelated to this pull request.

blowmage commented:
> In your gist you suggested class methods Schema::load and Schema::dump. Those are easy enough to implement, but while trying to document them, I realized that I have no idea when you would actually use that. Can you provide a suggestion? (It seems to me that Schema is always used within the context of a table in which case there's an instance of it already.)

My suggestion for `Schema.dump` and `Schema.load` is to match the interface used by [`Marshal.dump`](https://ruby-doc.org/core-2.5.0/Marshal.html#method-c-dump)/[`Marshal.load`](https://ruby-doc.org/core-2.5.0/Marshal.html#method-c-load), and [`JSON.dump`](http://ruby-doc.org/stdlib-2.5.0/libdoc/json/rdoc/JSON.html#method-i-dump)/[`JSON.load`](http://ruby-doc.org/stdlib-2.5.0/libdoc/json/rdoc/JSON.html#method-i-load). The Ruby idiom is that these are available as class methods.

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb:
```diff
@@ -384,6 +464,17 @@ def verify_mode mode
+        def migrate_json source_json
+          fields = JSON.parse source_json
+          {
+            "fields" => fields
```
Please use a symbol here to avoid unnecessary string creation.

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb:
```diff
@@ -542,6 +542,18 @@ def == other
+          # @private
+          def to_json *args
+            h = {
+              "name" => name,
```
Please use symbols here to avoid unnecessary string creation.

jeremywadsack commented on google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb:
```diff
@@ -123,6 +123,86 @@ def empty?
+
+          raise ArgumentError unless schema_json
+
+          @original_json = schema_json
```
It's set in [`::from_gapi`](https://github.com/keylimetoolbox/google-cloud-ruby/blob/a07c15fb43160b4ddae2617ddacd94c9ed116ba8/google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb#L407). I couldn't use that function directly, because I already had an instance I needed to change — I didn't want to create a new instance.

It looks like this is used by the [`#changed?` method](https://github.com/keylimetoolbox/google-cloud-ruby/blob/a07c15fb43160b4ddae2617ddacd94c9ed116ba8/google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb#L395).

Should I not be setting it here? Is this considered a change?

blowmage replied:
Ah, that makes sense now. If you call load on a schema should `changed?` be `true` until it is saved? @quartzmo, what do you think about this? What state should the schema object be in after calling `load`?

quartzmo replied:
It looks to me like calling `@original_json = schema_json` will result in `changed?` returning `false`, because `@original_json` will be the same as `@gapi.to_json`. If `changed?` is `false`, won't `Table#check_for_mutated_schema!` fail to persist the change? I might be wrong (still need to look for tests), but it seems to me this statement should be removed.

blowmage replied:
That’s my thinking. I don’t think `load` should be setting `@original_json`.

jeremywadsack replied:
Remove in fa42191

quartzmo replied:
The PR needs both acceptance and unit tests for using `load` within a block passed to `Table#schema` for update.

jeremywadsack commented:
@blowmage Code review addressed except for my open question about `@original_json`. The more I think about that the more I think it should be `changed?`. Just as if you were adding fields with any of the `add_fields` methods.

jeremywadsack commented:
I'm not sure what's going on with the CI failures. The test fail in the coveralls run (I think) but they pass in the unit test run. They _look_ like they are running the code from the penultimate commit, but the coveralls report has the right sha. When I run `rvm-exec 2.4.2 bundle exec rake test:coverage` locally for the `google-cloud-bigquery` folder I don't get any errors. Thoughts?

quartzmo commented on google-cloud-bigquery/acceptance/bigquery/table_test.rb:
```diff
@@ -191,6 +191,156 @@
+
   it "gets and sets time partitioning" do
     partitioned_table = dataset.table "weekly_kittens"
     if partitioned_table.nil?
```
We also need unit test(s) for `Table#schema` (add to `test/google/cloud/bigquery/table_schema_test.rb`) proving that loading a new schema results in a request to the service.

jeremywadsack replied:
"updates its schema" acceptance test added in a8069c6. Although it passed whether or not I set `@original_json` in the `Schema`.

quartzmo commented:
@jeremywadsack I had restarted the failed build just before you pushed your last commits. I cancelled my restart. `rake ci` passes for me locally. (It did before, too.)

quartzmo commented on google-cloud-bigquery/acceptance/bigquery/table_test.rb:
```diff
@@ -401,6 +401,32 @@
+      end
+      t.headers.must_equal %i[id breed name features]
+
+      t.schema replace: true do |s|
```
I don't think it will matter whether `replace` is `true` or `false` here, will it? It would be good to update the `Table#schema` docs to cover loading--with an example, and an explanation of how it will always be a complete replacement.

quartzmo replied:
Good questions! Can you add tests/assertions to see what happens, then document in `Table#schema` docs?

quartzmo commented:
The CI build has failed again with the same errors as before. I am unable to reproduce locally. `rake ci` passes for me.

```sh


  1) Error:

Google::Cloud::Bigquery::Schema::mock_bigquery#test_0008_can dump the schema as JSON to a filename:

NoMethodError: undefined method `[]' for nil:NilClass

    /home/ubuntu/google-cloud-ruby/google-cloud-bigquery/test/google/cloud/bigquery/schema_test.rb:577:in `block (2 levels) in <top (required)>'



  2) Error:

Google::Cloud::Bigquery::Schema::mock_bigquery#test_0007_can dump the schema as JSON to a File:

NoMethodError: undefined method `[]' for nil:NilClass

    /home/ubuntu/google-cloud-ruby/google-cloud-bigquery/test/google/cloud/bigquery/schema_test.rb:501:in `block (2 levels) in <top (required)>'



  3) Failure:

Google::Cloud::Bigquery::Schema::mock_bigquery#test_0010_can write write a schema with the class method [/home/ubuntu/google-cloud-ruby/google-cloud-bigquery/test/google/cloud/bigquery/schema_test.rb:661]:

--- expected

+++ actual

@@ -1 +1 @@

-["name", "age", "score", "active", "avatar", "started_at", "duration", "target_end", "birthday", "alts"]

+[nil, nil, nil, nil, nil, nil, nil, nil, nil, nil]
```

jeremywadsack commented:
I wonder if CI failure is related to changing keys from string to symbol in
the hash we use to generate json. It seems to have the right number of json
objects but wrong attributes.
On Tue, Apr 24, 2018 at 4:43 PM Chris Smith <notifications@github.com>
wrote:

> The CI build has failed again with the same errors as before. I am unable
> to reproduce locally. rake ci passes for me.
>
>   1) Error:
>
> Google::Cloud::Bigquery::Schema::mock_bigquery#test_0008_can dump the schema as JSON to a filename:
>
> NoMethodError: undefined method `[]' for nil:NilClass    /home/ubuntu/google-cloud-ruby/google-cloud-bigquery/test/google/cloud/bigquery/schema_test.rb:577:in `block (2 levels) in <top (required)>'  2) Error:Google::Cloud::Bigquery::Schema::mock_bigquery#test_0007_can dump the schema as JSON to a File:NoMethodError: undefined method `[]' for nil:NilClass    /home/ubuntu/google-cloud-ruby/google-cloud-bigquery/test/google/cloud/bigquery/schema_test.rb:501:in `block (2 levels) in <top (required)>'
>
>
>
>   3) Failure:
>
> Google::Cloud::Bigquery::Schema::mock_bigquery#test_0010_can write write a schema with the class method [/home/ubuntu/google-cloud-ruby/google-cloud-bigquery/test/google/cloud/bigquery/schema_test.rb:661]:
>
> --- expected
>
> +++ actual
>
> @@ -1 +1 @@
>
> -["name", "age", "score", "active", "avatar", "started_at", "duration", "target_end", "birthday", "alts"]
>
> +[nil, nil, nil, nil, nil, nil, nil, nil, nil, nil]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/2061#issuecomment-384114562>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOurLF7ZDxy1oGjC21cJe6n2I3O6jfZks5tr7izgaJpZM4TelgR>
> .
>
-- 
Jeremy Wadsack

quartzmo commented:
@blowmage Can you pull this branch and run `rake ci` locally to see if you can reproduce the error on Circle CI?

jeremywadsack commented:
@quartzmo I just pushed a change to move some of the hash keys back to strings. This should let us know if that's the cause of the failure because the root elements will work but any nested ones would still fail.

blowmage commented:
I've identified why the coverage CI task is failing. The coverage task loads and runs all libraries together, and other libraries have a dependency on ActiveSupport. ActiveSupport has their own protocol for JSON serialization, which looks like you are intending to support by supporting `*args` on the `to_json` method. But AS also expects to get primitive objects/non-JSON string by calling `as_json`. This functionality is provided by the different named method `json_hash`, so AS serializes the object instance variables instead. That causes the JSON structure to be different and the tests to fail.

One way to correct this is to use the naming that AS expects:

```diff
diff --git a/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb b/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb
index b817def07..7931b735d 100644
--- a/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb
+++ b/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb
@@ -544,18 +544,18 @@ module Google
 
           # @private
           def to_json *args
-            json_hash.to_json(*args)
+            as_json(*args).to_json
           end
 
           # @private
-          def json_hash
+          def as_json *args
             h = {
               name: name,
               type: type,
               mode: mode
             }
             h[:description] = description if description
-            h[:fields] = fields.map(&:json_hash) if record?
+            h[:fields] = fields.map { |field| field.as_json(*args) } if record?
             h
           end
```

I don't love this, to be honest. I don't like the use of `*args`, as we aren't actually implementing the JSON serialization options that AS expects. I think a far more idiomatic approach is to use standard conversion methods.

```diff
diff --git a/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb b/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb
index b817def07..23f625a65 100644
--- a/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb
+++ b/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb
@@ -543,19 +543,19 @@ module Google
           end
 
           # @private
-          def to_json *args
-            json_hash.to_json(*args)
+          def to_json
+            JSON.dump to_hash
           end
 
           # @private
-          def json_hash
+          def to_hash
             h = {
               name: name,
               type: type,
               mode: mode
             }
             h[:description] = description if description
-            h[:fields] = fields.map(&:json_hash) if record?
+            h[:fields] = fields.map(&:to_hash) if record?
             h
           end
```

Either approach will work, but I have a strong preference for the latter.

Regardless of which approach is chosen, we should also document the usage of the JSON library by adding the require to the file:

```diff
diff --git a/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb b/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb
index b817def07..82be4c0bd 100644
--- a/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb
+++ b/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb
@@ -13,6 +13,8 @@
 # limitations under the License.
 
 
+require "json"
+
 module Google
   module Cloud
     module Bigquery
```

quartzmo commented:
Great sleuthing, @blowmage! I agree that we should avoid implementing ActiveSupport's API *unless* there is a really good reason to do so (in which case, my understanding has always been to implement `as_json`). How does this sound, @jeremywadsack?

jeremywadsack commented:
@quartzmo This sounds great. @blowmage Thank you for figuring this out. I'm perfectly happy to implement the latter solution. I'll get that change later this afternoon or tomorrow.

jeremywadsack commented:
I implemented a mix of the two solutions in 9e12b7c, but this may still fail. I couldn't get `to_json` without arguments to work. It failed in tests with "wrong number of arguments given":

```
ArgumentError: wrong number of arguments (given 1, expected 0)
    google-cloud-ruby/google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb:547:in `to_json'
    google-cloud-ruby/google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb:258:in `to_json'
    google-cloud-ruby/google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb:258:in `dump'
    google-cloud-ruby/google-cloud-bigquery/test/google/cloud/bigquery/table_schema_test.rb:309:in `block (2 levels
```

Even though I'm not passing arguments when I call the `Fields#to_json` method, it was receiving a `JSON::Ext::Generator::State` object as a single argument. From what I can tell this is part of how the ruby JSON implementation works. I'll wait to see if that's still an issue for AS or not.

quartzmo commented:
Hooray, CI passed!

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb:
```diff
@@ -123,6 +184,85 @@ def empty?
+        #   table.schema.dump "path/to/schema.json"
+        #
+        def dump destination
+          if destination.respond_to?(:rewind) && destination.respond_to?(:write)
```
I bet if this code is changed to the following:

```ruby
if destination.respond_to?(:rewind) && destination.respond_to?(:write)
  destination.rewind
  destination.write JSON.dump(fields.map(&:to_hash))
else
  File.write String(destination), JSON.dump(fields.map(&:to_hash))
end
```

You can avoid the `*args` argument to `Field#to_json` (_and_ you can remove the `Field#to_json` method).

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb:
```diff
@@ -123,6 +184,85 @@ def empty?
+            source.rewind
+            schema_json = String source.read
+          elsif source.is_a? Array
+            schema_json = source.to_json
```
```ruby
schema_json = JSON.dump source
```

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb:
```diff
@@ -123,6 +184,85 @@ def empty?
+            schema_json = String source
+          end
+
+          schema_json = migrate_json schema_json
```
The more I think about this, the more I would leave it up to the `Google::Apis::BigqueryV2::TableSchema.from_json` method to raise an appropriate error. I think we could remove the `migrate_json` method completely and have this instead:

```ruby
schema_json = "{\"fields\":#{schema_json}}"

@gapi = Google::Apis::BigqueryV2::TableSchema.from_json schema_json

self
```

jeremywadsack replied:
Yeah, I thought about that too. I think this is a good suggestion.

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb:
```diff
@@ -219,13 +219,12 @@ def load source
           elsif source.is_a? Array
-            schema_json = source.to_json
+            schema_json = JSON.dump source
           else
```
Now that `JSON.dump` is being called here, the require should be moved to this file.

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb:
```diff
@@ -123,6 +184,84 @@ def empty?
+          end
+
+          schema_json = %({"fields":#{schema_json}})
+          raise ArgumentError, "Invalid JSON" unless schema_json
```
This `raise` can be removed. It will never be `nil` because of the previous line.

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/schema/field.rb:
```diff
@@ -12,7 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-
```
We need two newlines or YARD will use the copyright header as documentation for the `Google` module. Can you restore the newline?

quartzmo commented on google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb:
```diff
@@ -44,6 +45,67 @@ module Bigquery
+          # consisting of an array of JSON objects containing the following:
+          # - `name`: The column's [data
+          #   type](https://cloud.google.com/bigquery/docs/schemas#standard_sql_data_types)
+          # - `type`: The column [name](https://cloud.google.com/bigquery/docs/schemas#column_names)
```
Are the descriptions for `name` and `type` reversed here?

quartzmo commented on google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb:
```diff
@@ -123,6 +185,83 @@ def empty?
+        # consisting of an array of JSON objects containing the following:
+        # - `name`: The column's [data
+        #   type](https://cloud.google.com/bigquery/docs/schemas#standard_sql_data_types)
+        # - `type`: The column [name](https://cloud.google.com/bigquery/docs/schemas#column_names)
```
Are the descriptions for name and type reversed here?

quartzmo commented on google-cloud-bigquery/support/doctest_helper.rb:
```diff
@@ -596,6 +602,10 @@ def other_dataset_view_object
   end
 
+  doctest.before "Google::Cloud::Bigquery::Schema.load" do
+    skip "This reads a File object, which is difficult to mock with doctest."
```
If this is still true, do we need the `File` monkey-patch above?

jeremywadsack commented:
@blowmage Thanks. I still have several test cases to cover for @quartzmo that are on my task list when I get time to look at it again. I hope to wrap that up this week.

quartzmo commented:
@jeremywadsack Need any help with this one? I need to release another BigQuery feature soon (PR #2069), it would be convenient to release this at the same time if possible.

jeremywadsack commented:
Sorry @quartzmo won't be able to wrap that up in the short term. I need to
get my head around how BigQuery handles updates in order to finish the
tests. Really crushed right now; hopefully I can look at this next week.

If you have something else to release, I'd say don't wait on me. Sorry.
On Fri, May 4, 2018 at 12:03 PM Chris Smith <notifications@github.com>
wrote:

> @jeremywadsack <https://github.com/jeremywadsack> Need any help with this
> one? I need to release another BigQuery feature soon (PR #2069
> <https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/2069>), it
> would be convenient to release this at the same time if possible.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/2061#issuecomment-386664891>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOurABqY0AVyLf3L11frKDajke7pD0zks5tvIn0gaJpZM4TelgR>
> .
>
-- 
Jeremy Wadsack

quartzmo commented:
@jeremywadsack We have another BIgQuery release approaching for #2081. How's your bandwidth?

jeremywadsack commented:
@quartzmo The new commits include the unit tests that you asked for and fix the documentation. I was able to remove the `File` monkey patch by using (and removing) real files for the doctests. 

Let me know if there's anything I missed or if I need to update any tests or documentation. Once this is approved, I can squash and rebase (if needed).

blowmage commented:
This file should be in acceptance, not test.

jeremywadsack commented:
It's used for the test suite not the acceptance suite. Would you rather I
try to mock the file read for the unit test?
On Thu, May 17, 2018 at 7:44 AM Mike Moore <notifications@github.com> wrote:

> This file should be in acceptance, not test.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/2061#issuecomment-389892029>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOurGQbcSsZb9qQsaUroduN0lwn4lD_ks5tzYzbgaJpZM4TelgR>
> .
>
-- 
Jeremy Wadsack

quartzmo commented:
@jeremywadsack I'd definitely prefer to avoid using a real file in both doctests and unit tests. We consider monkey-patching acceptable for doctests. For unit tests, can you borrow from [this `File.stub` usage](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-bigquery/test/google/cloud/bigquery_test.rb#L110)?

jeremywadsack commented:
@quartzmo Ok, I'll change that back. I thought you had [asked me to remove the monkey patch](https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/2061#discussion_r184480092)  in the doctests. I misunderstood.

Thanks for the link to the `File.stub` example. That will help for the unit tests. 

We can't stub that in doctest (because there's no block and no `doctest.around`) and we determined that we couldn't monkey patch `File.read` because it's required to run tests. So I think that means we have to skip doctest for any test that reads a file.

quartzmo commented:
Yes, sorry for the misunderstanding. I was wondering: If the doctest is skipped, do we still need the monkeypatch? Apologies, this does get a bit confusing.

jeremywadsack commented:
@quartzmo Yeah I can skip the tests that depend on the monkey patch you'd rather go that route. You're call.

jeremywadsack commented:
@quartzmo @blowmage I changed the doctests to the monkey patch and skip any examples that would read from a file. For unit tests I removed that `schema` file and rewrote the tests I had to use `StringIO` instead of `File` for IO handling. I think that addresses your concerns. However, when I I did a quick check I saw that _lots_ of existing unit tests read files from `acceptance/data/`. Let me know if this is what you want here or if you need any more changes.

quartzmo commented:
@jeremywadsack Thank you! I've pulled your updates and will take a look this weekend. If we can approve and merge it, it will be released very soon!

quartzmo commented:
Acceptance tests failed on CI on all nodes with:

```sh
  1) Failure:

Google::Cloud::Bigquery::Schema::mock_bigquery#test_0010_can dump a schema with the class method [/home/ubuntu/google-cloud-ruby/google-cloud-bigquery/test/google/cloud/bigquery/schema_test.rb:691]:

Expected: 10

  Actual: 11



  2) Failure:

Google::Cloud::Bigquery::Schema::mock_bigquery#test_0008_can dump the schema as JSON to a filename [/home/ubuntu/google-cloud-ruby/google-cloud-bigquery/test/google/cloud/bigquery/schema_test.rb:607]:

Expected: 10

  Actual: 11



  3) Failure:

Google::Cloud::Bigquery::Schema::mock_bigquery#test_0007_can dump the schema as JSON to a File [/home/ubuntu/google-cloud-ruby/google-cloud-bigquery/test/google/cloud/bigquery/schema_test.rb:531]:

Expected: 10

  Actual: 11
```

Investigating, and I will push a fix when I figure it out.

blowmage commented:
This is likely because we added support for NUMERIC while this PR was being approved. The default schema now includes a NUMERIC field.
